projects/dialogue_unlikelihood/agents.py

```diff
@@ -0,0 +1,619 @@
+        n = self.opt['seq_ul_n']  # label n-grams
+        crep_mask = torch.zeros_like(pred_toks).type_as(logits)
+        lrep_mask = torch.zeros_like(pred_toks).type_as(logits)
+
+        mask_len = []
+        for i, gen in enumerate(generations):
+            gen_i = gen.tolist()
+
+            # Collect context ngrams
+            context_i = batch.text_vec[i].tolist()
+            context_n_grams = self._count_n_grams(context_i, n)
+
+            # Get corresponding label
+            label_i = batch.label_vec[i].tolist()
+
+            # Get lengths
+            len_human_label = self._get_length(label_i)
+            len_model_label = self._get_length(gen_i)
+
+            seen_n_grams = defaultdict(int)
+            seen_n_grams_count = 0
+
+            # penalize if there is a context repeat
+            for j, n_gram in enumerate(NGramIterator(gen_i, n)):
+                if context_n_grams[n_gram] > 0:
+                    crep_mask[i, j : j + n] = 1
+
+            # penalize if there is a label repeat
+            for j, n_gram in enumerate(NGramIterator(gen_i, n)):
+                if seen_n_grams[n_gram] > 0:
+                    lrep_mask[i, j : j + n] = 1
+                seen_n_grams[n_gram] += 1
+
+        # Compute unlikelihood loss
+        lprobs = self.pred_logsoftmax(logits)
+        pred_lprobs = lprobs.view(-1, lprobs.size(2)).gather(1, pred_toks.view(-1, 1))
+        one_minus_probs = torch.clamp((1.0 - pred_lprobs.exp()), min=clamp_min).view(
+            pred_toks.size(0), pred_toks.size(1)
+        )
+
+        mask = ((1 - self.opt['ctxt_beta']) * lrep_mask) + (
+            self.opt['ctxt_beta'] * crep_mask
+        )
+
+        ul_loss = -torch.log(one_minus_probs) * mask
+        total_loss = div(ul_loss.sum(), mask.sum())
+        self.record_local_metric(
+            'ul_loss', AverageMetric.many(ul_loss.sum(dim=-1), mask.sum(dim=-1))
+        )
+
+        if not self.is_training:
+            # in eval mode, we want metrics (e.g. PPL) provided by tga's compute_loss
+            _, _ = super().compute_loss(batch, return_output=True)
+
+        if return_output:
+            return total_loss, model_output
+        return total_loss
+
+    def eval_step(self, batch):
+        """Evaluate a single batch of examples."""
+        if batch.text_vec is None:
+            return
+        bsz = batch.text_vec.size(0)
+        self.model.eval()
+        cand_scores = None
+
+        if batch.label_vec is not None:
+            # calculate loss on targets with teacher forcing
+            _ = self.compute_loss(batch)  # noqa: F841  we need the side effects
+
+        maxlen = self.label_truncate or 256
+        beam_preds_scores, _ = self._generate(batch, self.beam_size, maxlen)
+        preds, scores = zip(*beam_preds_scores)
+
+        self._ngram_metrics(batch, preds)
+
+        cand_choices = None
+        # TODO: abstract out the scoring here?
+        if self.rank_candidates:
```

Comment:
would be nice if this was abstracted out in TGA so that we avoid the copypasta here